lib/sshkit.ex

```diff
@@ -262,17 +263,96 @@ defmodule SSHKit do
     Enum.map(context.hosts, run)
   end
 
-  # def upload(context, path, options \\ []) do
-  #   …
-  #   # resolve remote relative to context path
-  #   remote = Path.expand(Map.get(options, :as, Path.basename(path)), _)
-  #   SCP.upload(conn, path, remote, options)
-  # end
-
-  # def download(context, path, options \\ []) do
-  #   …
-  #   remote = _ # resolve remote relative to context path
-  #   local = Map.get(options, :as, Path.basename(path))
-  #   SCP.download(conn, remote, local, options)
-  # end
+  @doc ~S"""
+  Upload a file or files to the given context.
+  Returns a list of `:ok`.
+  There is one `:ok` per connected host where file upload was successful.
+
+  ## Examples
+
+  Upload all files and folders in current directory to "/workspace".
+
+  ```
+  [:ok] =
+    "my.remote-host.tld"
+    |> SSHKit.context
+    |> SSHKit.path("/workspace")
+    |> SSHKit.upload(".", recursive: true)
+
+  ```
+
+  Upload file to different name on host
+
+  ```
+  [:ok] =
+    "my.remote-host.tld"
+    |> SSHKit.context
+    |> SSHKit.upload("some_file.txt", as: "other_file.txt" )
+
+  ```
+
+  """
+  def upload(context, path, options \\ []) do
+    # resolve remote relative to context path
+    # override with :as option if present
+    destination_path = if (context.path), do: context.path, else: ""
+    destination = case Keyword.get(options, :as, false) do
+      false -> path
+      override -> override
+    end
+    remote = if (Path.type(destination) == :absolute), do: destination, else: Path.join(destination_path, destination)
+
+    run = fn host ->
+      {:ok, conn} = SSH.connect(host.name, host.options)
+      SCP.upload(conn, path, remote, options)
+    end
+
+    Enum.map(context.hosts, run)
+  end
```

Comment:
Sure that's outta the scope of this PR. I'm pretty new to the project, but one should probably put some thought into how various functions from the DSL will be combined and based on that decide what would make sense as return values.

I could for instance see the following approaches:

**a) Return `{:ok, conn, _x}`**
This would allow chaining commands like this:
```elixir
with {:ok, conn}      <- SSHKit.SSH.connect("foo.io"),
     {:ok, conn, :ok} <- SSHKit.SCP.upload(conn, "local", "remote"),
     {:ok, conn, :ok} <- SSHKit.SCP.upload(conn, "other_local", "remote"),
     {:ok, conn, res} <- SSHKit.SSH.run(conn, "command"),
     {:ok, conn, :ok} <- SSHKit.SSH.close(conn)
do
     {:ok, res}
else
     err -> err
end
```

**b) Return a connection with certain fields set**
This would be similar to how [Plug.Conn](https://hexdocs.pm/plug/Plug.Conn.html#content) works afaik. Building workflows could then look like this:
```elixir
"foo.io"
|> SSHKit.SSH.connect()
|> SSHKit.SCP.upload("local", "remote")
|> SSHKit.SSH.run("command")
|> SSHKit.SSH.close()
```

I'm personally more in favor of building workflows with "with", since it allows a better / more fine granular control over how errors are handled, but that comes down to personal preference and the general vision of the DSL I guess.